File: ios-app/Sources/RedCodeFeatures/E2eeSessionLifecycle.swift
import Combine
import Foundation
import RedCodeCore
import RedCodeNetworking

public enum E2eeSessionStatus: Equatable, Sendable {
    case signedOut
    case plaintext
    case ready(accountID: String, deviceID: String)
    case blocked(message: String)
}

public protocol E2eeRuntimeSettingsProviding: Sendable {
    func fetchMessageRuntime() async throws -> MessageRuntimeSettings
}

public struct SettingsE2eeRuntimeProvider: E2eeRuntimeSettingsProviding {
    private let settings: any SettingsAPIService

    public init(settings: any SettingsAPIService) {
        self.settings = settings
    }

    public func fetchMessageRuntime() async throws -> MessageRuntimeSettings {
        try await settings.fetchGeneralSettings().messageRuntime
    }
}

public protocol E2eeAppDeviceLifecycle: Sendable {
    func ensureReady(accountID: String, deviceLabel: String, token: String) async throws -> E2eeDeviceProfile
    func topUpKeyPackages(accountID: String, token: String) async throws -> Int
}

extension E2eeDeviceLifecycle: E2eeAppDeviceLifecycle {}

public protocol E2eeAccountSecureStateClearing: Sendable {
    func delete(accountID: String) async throws
}

public struct E2eeAccountSecureStateCleaner: E2eeAccountSecureStateClearing {
    private let operation: @Sendable (String) async throws -> Void

    public init(operation: @escaping @Sendable (String) async throws -> Void) {
        self.operation = operation
    }

    public func delete(accountID: String) async throws {
        try await operation(accountID)
    }
}

public struct E2eeSessionLifecycleError: Error, Equatable, LocalizedError, Sendable {
    public let message: String

    public init(_ message: String) {
        self.message = message
    }

    public var errorDescription: String? { message }
}

@MainActor
public final class E2eeSessionLifecycle: ObservableObject {
    @Published public private(set) var status: E2eeSessionStatus = .signedOut

    private struct ActiveSession: Equatable, Sendable {
        let accountID: String
        let token: String
    }

    private let settings: any E2eeRuntimeSettingsProviding
    private let devices: any E2eeAppDeviceLifecycle
    private let secureState: any E2eeAccountSecureStateClearing
    private let deviceLabel: String
    private var activeSession: ActiveSession?
    private var preparationTask: Task<E2eeSessionStatus, Error>?
    private var preparationID: UUID?

    public init(
        settings: any E2eeRuntimeSettingsProviding,
        devices: any E2eeAppDeviceLifecycle,
        secureState: any E2eeAccountSecureStateClearing,
        deviceLabel: String
    ) {
        self.settings = settings
        self.devices = devices
        self.secureState = secureState
        self.deviceLabel = deviceLabel
    }

    public func onAuthenticated(session: AuthSession) async throws {
        let accountID = session.user.id.trimmingCharacters(in: .whitespacesAndNewlines)
        let token = session.token.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !accountID.isEmpty else {
            throw E2eeSessionLifecycleError("E2EE 账号标识不能为空")
        }
        guard !token.isEmpty else {
            throw E2eeSessionLifecycleError("E2EE 认证 token 不能为空")
        }

        if let previous = activeSession, previous.accountID != accountID {
            cancelPreparation()
            activeSession = nil
            do {
                try await secureState.delete(accountID: previous.accountID)
            } catch {
                status = .blocked(message: Self.message(for: error))
                throw error
            }
        }
        let current = ActiveSession(accountID: accountID, token: token)
        activeSession = current
        try await refreshAndPrepare(current)
    }

    public func onForeground() async throws {
        guard let activeSession else { return }
        try await refreshAndPrepare(activeSession)
    }

    public func onLogout() async throws {
        let accountID = activeSession?.accountID
        activeSession = nil
        cancelPreparation()
        status = .signedOut
        if let accountID {
            try await secureState.delete(accountID: accountID)
        }
    }

    public func requireE2eeReady() throws -> (accountID: String, deviceID: String) {
        guard case .ready(let accountID, let deviceID) = status else {
            let message = if case .blocked(let message) = status {
                message
            } else {
                "当前设备未进入 E2EE ready 状态"
            }
            throw E2eeSessionLifecycleError(message)
        }
        return (accountID, deviceID)
    }

    private func refreshAndPrepare(_ session: ActiveSession) async throws {
        if let preparationTask {
            try await apply(preparationTask, to: session)
            return
        }

        let taskID = UUID()
        let settings = self.settings
        let devices = self.devices
        let deviceLabel = self.deviceLabel
        let task = Task<E2eeSessionStatus, Error> {
            let runtime = try await settings.fetchMessageRuntime()
            try Self.validate(runtime)
            guard runtime.isE2EE else { return .plaintext }

            let profile = try await devices.ensureReady(
                accountID: session.accountID,
                deviceLabel: deviceLabel,
                token: session.token
            )
            guard profile.deviceStatus == "active" else {
                throw E2eeSessionLifecycleError(
                    "E2EE 设备状态为 \(profile.deviceStatus)，拒绝进入加密消息链"
                )
            }
            _ = try await devices.topUpKeyPackages(accountID: session.accountID, token: session.token)
            return .ready(accountID: session.accountID, deviceID: profile.deviceId)
        }
        preparationTask = task
        preparationID = taskID
        defer {
            if preparationID == taskID {
                preparationTask = nil
                preparationID = nil
            }
        }
        try await apply(task, to: session)
    }

    private func apply(_ task: Task<E2eeSessionStatus, Error>, to session: ActiveSession) async throws {
        do {
            let nextStatus = try await task.value
            guard activeSession == session else { return }
            status = nextStatus
        } catch {
            guard activeSession == session else { throw error }
            status = .blocked(message: Self.message(for: error))
            throw error
        }
    }

    private func cancelPreparation() {
        preparationTask?.cancel()
        preparationTask = nil
        preparationID = nil
    }

    private static func validate(_ runtime: MessageRuntimeSettings) throws {
        guard ["persist", "relay_only"].contains(runtime.serverStorageMode) else {
            throw E2eeSessionLifecycleError("未知的消息存储模式：\(runtime.serverStorageMode)")
        }
        guard ["plaintext", "e2ee"].contains(runtime.contentAuditMode) else {
            throw E2eeSessionLifecycleError("未知的消息审计模式：\(runtime.contentAuditMode)")
        }
    }

    private static func message(for error: Error) -> String {
        if let error = error as? E2eeSessionLifecycleError { return error.message }
        if let error = error as? E2eeDeviceNotReadyError { return error.message }
        return error.localizedDescription
    }
}

File: android-app/app/src/main/java/com/redcode/im/androidapp/e2ee/E2eeSessionLifecycle.kt
package com.redcode.im.androidapp.e2ee

import com.redcode.im.androidapp.core.model.MessageRuntimeSettings
import com.redcode.im.androidapp.data.settings.SettingsRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

sealed interface E2eeSessionStatus {
    data object SignedOut : E2eeSessionStatus
    data object Plaintext : E2eeSessionStatus
    data class Ready(val accountId: String, val deviceId: String) : E2eeSessionStatus
    data class Blocked(val message: String) : E2eeSessionStatus
}

interface E2eeAppDeviceLifecycle {
    suspend fun ensureReady(accountId: String, deviceLabel: String, token: String): E2eeDeviceProfile

    suspend fun topUpKeyPackages(accountId: String, token: String): Int
}

/** 账号级应用生命周期入口。未知 runtime 或设备异常会保留 Blocked 状态，供消息路由 fail closed。 */
class E2eeSessionLifecycle(
    private val settings: SettingsRepository,
    private val devices: E2eeAppDeviceLifecycle,
    private val secureState: E2eeSecureStateStore,
    private val deviceLabel: String,
) {
    private data class ActiveSession(val accountId: String, val token: String)

    private val mutex = Mutex()
    private val mutableStatus = MutableStateFlow<E2eeSessionStatus>(E2eeSessionStatus.SignedOut)
    val status = mutableStatus.asStateFlow()
    private var activeSession: ActiveSession? = null

    suspend fun onAuthenticated(accountId: String, token: String) = mutex.withLock {
        require(accountId.isNotBlank()) { "E2EE 账号标识不能为空" }
        require(token.isNotBlank()) { "E2EE 认证 token 不能为空" }
        activeSession
            ?.accountId
            ?.takeIf { it != accountId }
            ?.let(secureState::delete)
        activeSession = ActiveSession(accountId, token)
        refreshAndPrepare(activeSession!!)
    }

    suspend fun onForeground() = mutex.withLock {
        activeSession?.let { refreshAndPrepare(it) }
    }

    suspend fun onLogout() = mutex.withLock {
        val accountId = activeSession?.accountId
        activeSession = null
        if (accountId != null) secureState.delete(accountId)
        mutableStatus.value = E2eeSessionStatus.SignedOut
    }

    fun requireE2eeReady(): E2eeSessionStatus.Ready =
        status.value as? E2eeSessionStatus.Ready
            ?: throw E2eeDeviceNotReadyException(
                (status.value as? E2eeSessionStatus.Blocked)?.message ?: "当前设备未进入 E2EE ready 状态",
            )

    private suspend fun refreshAndPrepare(session: ActiveSession) {
        try {
            val runtime = settings.refreshGeneralSettings().messageRuntime.requireSupported()
            if (!runtime.isE2ee) {
                mutableStatus.value = E2eeSessionStatus.Plaintext
                return
            }
            val profile = devices.ensureReady(session.accountId, deviceLabel, session.token)
            if (profile.deviceStatus != "active") {
                throw E2eeDeviceNotReadyException("E2EE 设备状态为 ${profile.deviceStatus}，拒绝进入加密消息链")
            }
            devices.topUpKeyPackages(session.accountId, session.token)
            mutableStatus.value = E2eeSessionStatus.Ready(session.accountId, profile.deviceId)
        } catch (error: Exception) {
            mutableStatus.value = E2eeSessionStatus.Blocked(error.message ?: "E2EE 初始化失败")
            throw error
        }
    }
}

internal fun MessageRuntimeSettings.requireE2ee(): MessageRuntimeSettings =
    requireSupported().also { require(it.isE2ee) { "当前 message runtime 不是 E2EE" } }
